src/layers/normalization.h

```diff
@@ -172,4 +172,67 @@ class RMSNormImpl : public torch::nn::Module {
 };
 TORCH_MODULE(RMSNorm);
 
+// Root mean square normalization
+class RMSNormResidualImpl : public torch::nn::Module {
+ public:
+  RMSNormResidualImpl(int64_t dim,
+              float eps,
+              torch::ScalarType dtype,
+              const torch::Device& device)
+      : eps_(eps) {
+    weight_ = register_parameter(
+        "weight",
+        torch::empty({dim}, torch::dtype(dtype).device(device)),
+        /*requires_grad=*/false);
+  }
+
+  torch::Tensor forward(torch::Tensor input, torch::Tensor* residual) {
+    if (input.is_cuda() && !FLAGS_disable_custom_kernels) {
+      auto output = torch::empty_like(input);
+      if (residual != nullptr) {
+        input = input + *residual;
+        *residual = input;
+      }
+      kernel::rms_norm(output, input, weight_, eps_);
+      return output;
+    }
+    if (residual != nullptr) {
```

Comment:
It appears that introducing this class with a residual might be optional since this logic could be entirely moved out of the function. For instance:
```
if (residual.defined()) {
    input = input + residual;
    residual = input;
}
rms_norm(input);
```

but it may be necessary to be a new class if we want to fuse the kernel in the future.

Reply:
yes, a new class with fused kernel could be better.